### PR TITLE
[inductor] Fix mix-order reduction dropping writes to NonOwningLayout outputs

### LINKED ISSUE #193061: `torch.compile` silently computes wrong gradients for chained modulated LayerNorms with a shared scale/shift parameter (regression 2.9.1 → 2.10.0 due to mix_order_reduction)
### 🐛 Describe the bug

When a model applies two chained "modulated norm" blocks of the form `x = x + (layer_norm(x) * (1 + scale) + shift)`, where all `scale`/`shift` vectors are **chunks of one shared parameter** (the DiT / AdaLN `scale_shift_table` pattern used by Wan 2.1), the Inductor-compiled backward produces **silently wrong gradients for the second norm's `shift`/`scale` slices** once the total reduction size `batch*seq` is large (≳16k tokens). The first norm's slices are correct to fp32 precision; the second norm's slices come back with relative errors of up to ~1e+6 — depending on graph arrangement the corresponding region of the grad buffer is simply never written, so it can also surface as bitwise-stale values from a previous backward.

**Update: This is a regression introduced in 2.10.0**: the same script is fully clean on 2.9.1 and fails identically on 2.10.0, 2.11.0, and 2.13.0 (latest stable).

Minimal repro in plain fp32, plain `nn.LayerNorm`, no autocast, no checkpointing, no attention:

```python
import torch
import torch.nn as nn

DIM, SEQ, SMALL_SEQ = 1024, 32768, 4096
dev = 'cuda'


def run(seq=SEQ, n_norms=2, slicing='chunk'):
    torch.manual_seed(0)
    norms = [nn.LayerNorm(DIM, 1e-6, elementwise_affine=False).to(dev)
             for _ in range(n_norms)]
    if slicing == 'separate':
        params = [nn.Parameter(torch.randn(DIM, device=dev) / DIM ** 0.5)
                  for _ in range(2 * n_norms)]
    else:
        table = nn.Parameter(
            torch.randn(1, 2 * n_norms, DIM, device=dev) / DIM ** 0.5)
        params = [table]

    def fn(x):
        if slicing == 'separate':
            mods = params
        elif slicing == 'unbind':
            mods = [s.unsqueeze(1) for s in params[0].unbind(1)]
        else:  # 'chunk' — the failing case
            mods = params[0].chunk(2 * n_norms, dim=1)
        for i, norm in enumerate(norms):
            shift, scale = mods[2 * i], mods[2 * i + 1]
            x = x + (norm(x) * (1 + scale) + shift)
        return x.square().mean()

    fn_c = torch.compile(fn, mode='default', dynamic=False)
    x = torch.randn(1, seq, DIM, device=dev)

    def grads():
        g = [p.grad for p in params]
        for p in params:
            p.grad = None
        return torch.cat([t.reshape(-1) for t in g])

    fn(x).backward()
    g_eager = grads()
    fn_c(x).backward()
    g_compiled = grads()

    rel = ((g_eager - g_compiled).abs().max()
           / g_eager.abs().max().clamp(min=1e-12)).item()
    return rel


CASES = [
    ('chunked table, 2 norms, seq 32768  (THE BUG)', dict()),
    ('same but seq 4096', dict(seq=SMALL_SEQ)),
    ('same but single norm', dict(n_norms=1)),
    ('same but unbind(1)+unsqueeze(1) instead of chunk', dict(slicing='unbind')),
    ('same but 4 independent parameters', dict(slicing='separate')),
]

for name, kwargs in CASES:
    rel = run(**kwargs)
    print(f'{"FAIL" if rel > 0.1 else "ok  "}  max rel grad err {rel:9.3e}  {name}',
          flush=True)

print('\nexpected: first row FAIL (silently wrong grads), all others ok')
```

### Output

On H100 — **identical on torch 2.10.0, 2.11.0 and 2.13.0** (cu126/cu128/cu129 builds all affected; CUDA toolkit version is irrelevant). **torch 2.9.1 is clean** (first row `ok  2.141e-07`):

```
FAIL  max rel grad err 1.481e+06  chunked table, 2 norms, seq 32768  (THE BUG)
ok    max rel grad err 2.068e-07  same but seq 4096
ok    max rel grad err 2.251e-07  same but single norm
ok    max rel grad err 2.141e-07  same but unbind(1)+unsqueeze(1) instead of chunk
ok    max rel grad err 2.116e-07  same but 4 independent parameters
```

### What we bisected (each ingredient is necessary)

1. **Two chained modulated norms** — a single one is clean.
2. **Gradient flow through a multiplicative rsqrt-variance statistic** — an RMSNorm-style `x * rsqrt(mean(x^2) + eps)` fails exactly like LayerNorm; removing the norm, subtracting only the mean (`x - x.mean(-1, keepdim=True)`), or detaching the norm statistics (same forward, pointwise backward) are all clean.
3. **The modulation vectors must be `chunk`/narrow slices of one shared tensor** — `unbind(1)` (+`unsqueeze(1)` for the broadcast shape) or independent parameters are clean. The shared tensor need not be a parameter: chunking a conditioning MLP's output fails the same way (max rel err 7.2e+1 / 4.7e+5 on 2.11 / 2.13, landing in the MLP weight grads). In every failing case the slice grads assemble into one buffer — matching the never-written/stale-buffer behavior.
4. **The slices must broadcast over batch AND sequence, with total `batch*seq` ≳ 16k** — at DIM=5120: seq 8192 ok, seq 16384 FAIL; also FAIL at DIM=256 with long seq, and at batch 8 × seq 4096 (shared table) even though batch 1 × seq 4096 is ok. Per-sample modulation (a real batch dim, e.g. `lin(c).view(B, 4, D)` with B ≥ 2) is clean at every tested length (B=2 × seq 32768, B=4 × seq 65536 ok) — but at B=1 it degenerates to the broadcast case and fails (B=1 × seq 32768 FAIL). Note the two roles of batch size: within the broadcast case it only contributes to the total (B=8 × 4096 fails like B=1 × 32768), while B ≥ 2 per-sample modulation is clean even with a 65536-long reduction per grad element — so the broadcast structure selects the buggy codegen path, and reduction size is the trigger only within it.

Not required: autocast/bf16, gradient checkpointing, attention/FFN, `elementwise_affine`. No escape via compile config (`default`/`reduce-overhead`/`max-autotune`, `fullgraph` on/off, `dynamic` on/off), `TORCHINDUCTOR_SPLIT_REDUCTIONS=0`, or expression rewrites (cloning the norm output or slices, `contiguous()`, `torch.addcmul`, distributing the multiply) — only the `unbind`/separate-parameter rewrites, which change how the slice grads are routed, avoid it.

### Real-world impact

Found while training diffusers Wan 2.1 14B (video diffusion): the per-block `scale_shift_table` gradients were garbage, and since the grad-buffer region is never written, iteration N reused iteration N-1's values bitwise whenever shapes matched — a linearly growing grad-norm staircase.

Who is exposed: any chained-modulated-norm model whose modulation slices broadcast over the batch — learnable scale-shift tables (Wan-style, at any batch size), or ordinary per-sample adaLN whenever the local batch is 1 (common in long-video training and gradient accumulation). Per-sample adaLN at local batch ≥ 2 tested clean at every length (up to seq 65536), and we verified three real DiT-family ImageNet training configs are unaffected.

### Versions

Fails on 2.10.0+cu128, 2.11.0+cu128/cu129, 2.13.0+cu126/cu129; clean on 2.9.1+cu128 (regression window: 2.9.1 → 2.10.0). `collect_env` for the primary environment:

```
PyTorch version: 2.13.0+cu129
Is debug build: False
CUDA used to build PyTorch: 12.9
OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Libc version: glibc-2.35
Python version: 3.11.10 (64-bit runtime)
Python platform: Linux-5.15.0-131-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
GPU models and configuration: 8x NVIDIA H100 80GB HBM3
Nvidia driver version: 550.127.08
cuDNN version: 9.10.2
triton: 3.7.1
```

### A related observation

On 2.11.0 only, the same setup with independent parameters of shape `[1, 1, D]` (instead of `[D]` used above) made the compiled backward crash outright (`Function CompiledFunctionBackward returned an invalid gradient at index 3 - got [1024] but expected shape compatible with [1, 1, 1024]`). 2.13.0 neither crashes nor miscomputes there — so the grad-shape handling in this area changed between 2.11 and 2.13 while the silent bug above survived; possibly a useful hint for locating it.

### AI assistance disclosure

Per the [PyTorch AI policy](https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md):
The bug was hit in a real training run (Wan 2.1 14B). The debugging, ablation, and this write-up were developed with AI assistance (Claude Fable 5). Every number and code path in this report was executed and verified on our cluster (H100, both torch versions), and I have reviewed and can speak to all of it.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @ezyang @gchanan @zou3519 @msaroufim @eellison @shunting314 @jansel @bdhirsh

### PR DESCRIPTION
**Pull request by haojiang01**
Fixes #193061

## Problem

`torch.compile` silently computes wrong gradients for chained modulated LayerNorms whose scale/shift vectors are `chunk()`-ed slices of one shared parameter (the DiT/AdaLN `scale_shift_table` pattern, e.g. Wan 2.1), once the total reduction size is large enough to trigger mix-order reduction. Regression introduced when `triton.mix_order_reduction` became default-enabled in 2.10.

## Root cause

The slice gradients are assembled into one shared buffer by `cat`, so each mix-order-fused reduction output is a view into that buffer's storage (`ir.NonOwningLayout`). In `_codegen_mix_order_reduction`, the final-reduce step emits

```python
buf3 = workspace_0[...].view(nsplit, rnumel).sum(dim=0).view(...)
```

which **rebinds the Python variable** to a fresh standalone tensor. The `buf3 = reinterpret_tensor(buf13, ..., 2048)  # alias` line emitted by `mark_run()` is discarded, the corresponding region of the shared buffer is never written, and the returned gradient contains uninitialized (or bitwise-stale, when the allocator reuses the block across iterations) memory.

Generated backward before the fix (`buf13` is the `[1, 4, D]` table grad; offsets 2048/3072 never written):

```python
buf3 = workspace_0[0 * (1024) * (1024) : (0 + 1) * (1024) * (1024)].view(1024, 1024).sum(dim=0).view([1, 1, 1024])
buf5 = workspace_0[1 * (1024) * (1024) : (1 + 1) * (1024) * (1024)].view(1024, 1024).sum(dim=0).view([1, 1, 1024])
...
buf13 = empty_strided_cuda((1, 4, 1024), (4096, 1024, 1), torch.float32)
buf10 = reinterpret_tensor(buf13, (1, 1, 1024), (4096, 1024, 1), 0)  # alias
buf12 = reinterpret_tensor(buf13, (1, 1, 1024), (4096, 1024, 1), 1024)  # alias
return (buf13, None, )
```

## Fix

When the reduction output buffer has `NonOwningLayout`, emit the allocation for the view (idempotent) and write through it with `copy_()` instead of rebinding the name:

```python
buf3 = reinterpret_tensor(buf13, (1, 1, 1024), (4096, 1024, 1), 2048)  # alias
buf3.copy_(workspace_0[...].view(1024, 1024).sum(dim=0).view([1, 1, 1024]))
```

Non-view outputs keep the existing rebinding path unchanged.

## Test plan

- New regression test `test_chained_modulated_norm_shared_table` (parametrized over `split_reductions`) in `test/inductor/test_mix_order_reduction.py` — fails before the fix (max rel grad err ~1e+6), passes after (~1e-7).
- Full repro script from #193061 on H20: all 5 cases pass after the fix; also verified the issue's ablations (batch 8 × seq 4096, dim 5120 × seq 16384, and chunking a conditioning MLP's output instead of a parameter).
- `python -m pytest test/inductor/test_mix_order_reduction.py`: 277 passed, 221 skipped, 0 failed.
- `lintrunner` clean on changed files.

Per the [PyTorch AI policy](https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md): this PR was authored with the assistance of an AI coding assistant. The root-cause analysis, the change, and all test results were reviewed and verified by the submitter.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @shunting314 @eellison @jansel @vkuzo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by haojiang01** [comment]
Addressed all four review items in the latest commit:

1. **Explicit spec dispatch with loud fallback**: `NonOwningLayout` → write through the view; only exact `FixedLayout`/`FlexibleLayout` keep the rebinding path (exact `type()` check so `CommBufferLayout`, a `FixedLayout` subclass that must not be rebound, falls through); anything else raises `AssertionError` naming the buffer and spec. Also skip buffers in `V.graph.removed_buffers` to avoid the potential `NameError` after `codegen_allocation` emits nothing.
2. **Dead guards**: replaced both with `if not isinstance(buffer, ir.Buffer): raise AssertionError(type(buffer))` — loud instead of silently routing to the rebinding branch, and still satisfies the type checker.
3. **Redundant cast**: `.to(dtype)` is now appended only on the rebinding path; `copy_()` handles the conversion itself.
4. **Test**: the backward wrapper is captured via `utils.run_and_get_code`; `FileCheck` asserts the `.copy_(workspace_` write-through, `assertNotIn("= workspace_", wrapper)` rejects any rebinding from the workspace, and the metric check now uses `assertEqual(metrics.codegen_mix_order_reduction, 1)`.

Re-verified on H20: the issue's full repro (all 5 cases ≤ 2.3e-7), the strengthened test (both `split_reductions` variants), and the whole `test_mix_order_reduction.py` suite (279 passed, 0 failed). `lintrunner` clean.

**Comment by jansel** [comment]
@claude review this again

**Review by jansel — APPROVED** [review]

**Comment by haojiang01** [comment]
Note for reviewers: both CI failures on this PR are infra / pre-existing, not caused by this change.

**1. `linux-jammy-cuda13.2-cudnn9-py3.10-clang20 / build`** — [job](https://github.com/pytorch/pytorch/actions/runs/31658408353/job/94371566546)

Failed in `Initialize containers`, before any build step ran. The ARC runner pod could not pull the CI image and looped on `ErrImagePull` / `ImagePullBackOff` until the 7200s backoff timeout:

```
Pod ...-workflow: phase=Pending, ... waiting=[job: ImagePullBackOff] (7205s/7200s)
##[error]Error: pod failed to come online with error: Error: Pod ... is unhealthy with phase status Pending: backoff timeout
```

Image: `.../pytorch/ci-image:pytorch-linux-jammy-cuda13.2-cudnn9-py3.10-clang20-e480c7d79ae60f1b71fcbf47213d809666e9d33f`

**2. `torchtitan-x-pytorch-test / test (torchtitan_models_integration)`** — [job](https://github.com/pytorch/pytorch/actions/runs/32159571574/job/95788280000)

Four flavors fail with `ImportError` raised from `torchtitan/config/manager.py:120`, i.e. the test matrix references model modules that do not exist in the pinned torchtitan checkout:

| flavor | rc | error |
| --- | --- | --- |
| `qwen3_5_fsdp+tp+varlen_attn+per_op_sac` | 1 | `Cannot import config_registry for module 'qwen3_5'` |
| `muse_glimmer_mm_fsdp+tp+sp` | 1 | `Cannot import config_registry for module 'muse_glimmer'` |
| `kimi_k2_5_mm_fsdp_spmd_types` | 1 | `Cannot import config_registry for module 'kimi_k2_7'` |
| `kimi_k2_5_mm_fsdp+tp+ep+pp` | 1 | `Cannot import config_registry for module 'kimi_k2_7'` |

This is a broken trunk failure — the exact same four flavors fail with the exact same errors on `main`: [run 32178802263, job 95860007046](https://github.com/pytorch/pytorch/actions/runs/32178802263/job/95860007046) (sha `7b260c05`).

So `@pytorchbot merge -i` should be safe here once the review is done. Happy to rebase again if that helps pick up a fixed torchtitan pin.
